### ISSUE TITLE
NPE in JavascriptClientCodegen if definition name does not start with an upper case character

### ISSUE THREAD

**Issue by demonfiddler** [swagger-api/swagger-codegen#2146]
JavaScript code generation fails with NullPointerException under some, but not all, cases where a definition does not start with an upper case character. The problem does not arise for petstore-simple, but it does arise for our proprietary YAML which uses the allOf construct to implement inheritance. My employer's IP policy prohibits me from posting the offending YAML.

From what I've seen in the debugger this is because JavascriptClientCodegen line 404 makes the unwarranted assumption that toModelName(codegenModel.parent) will yield an existing key in allDefinitions. If a definition name starts with a lower case character this assumption will fail and parentModel will be passed as null.

I'm afraid I don't know enough about the code to suggest a fix.
```
Thread [main] (Suspended (breakpoint at line 404 in JavascriptClientCodegen))	
	JavascriptClientCodegen.fromModel(String, Model, Map<String,Model>) line: 404	
	DefaultGenerator.processModels(CodegenConfig, Map<String,Model>, Map<String,Model>) line: 750	
	DefaultGenerator.generate() line: 196	
	CodeGenMojo.execute() line: 207	
	DefaultBuildPluginManager.executeMojo(MavenSession, MojoExecution) line: 134	
	MojoExecutor.execute(MavenSession, MojoExecution, ProjectIndex, DependencyContext) line: 208	
	MojoExecutor.execute(MavenSession, MojoExecution, ProjectIndex, DependencyContext, PhaseRecorder) line: 153	
	MojoExecutor.execute(MavenSession, List<MojoExecution>, ProjectIndex) line: 145	
	LifecycleModuleBuilder.buildProject(MavenSession, MavenSession, ReactorContext, MavenProject, TaskSegment) line: 116	
	LifecycleModuleBuilder.buildProject(MavenSession, ReactorContext, MavenProject, TaskSegment) line: 80	
	SingleThreadedBuilder.build(MavenSession, ReactorContext, ProjectBuildList, List<TaskSegment>, ReactorBuildStatus) line: 51	
	LifecycleStarter.execute(MavenSession) line: 128	
	DefaultMaven.doExecute(MavenExecutionRequest, MavenSession, MavenExecutionResult, DefaultRepositorySystemSession) line: 307	
	DefaultMaven.doExecute(MavenExecutionRequest) line: 193	
	DefaultMaven.execute(MavenExecutionRequest) line: 106	
	MavenCli.execute(CliRequest) line: 862	
	MavenCli.doMain(CliRequest) line: 286	
	MavenCli.main(String[], ClassWorld) line: 197	
	NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
	NativeMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	Method.invoke(Object, Object...) line: not available	
	Launcher.launchEnhanced(String[]) line: 289	
	Launcher.launch(String[]) line: 229	
	Launcher.mainWithExitCode(String[]) line: 415	
	Launcher.main(String[]) line: 356	
```

**Comment by wing328** [comment]
@demonfiddler can you repeat the issue by updating https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/test/resources/2_0/petstore.json ?

**Comment by demonfiddler** [comment]
Sadly, no @wing328. I just tried lowercasing the first character of all petstore YAML definition names and $refs thereto, but codegen succeeded. As mentioned above I suspect it's related to our use of the allOf construct to implement inheritance.

**Comment by wing328** [comment]
I don't think allOf is supported in JS or other codegen at the moment...

**Comment by demonfiddler** [comment]
Are you quite sure? As I said, codegen from the offending YAML (which uses allOf) can be made to work by manually editing lowerCamelCase definition names and references to TitleCase. At any rate, this expedient avoids the NPE - as for whether the generated API actually works, I've yet to determine.

**Comment by demonfiddler** [comment]
Confirmed: allOf is not supported. I've logged Issue #2161 "allOf ignored, composition impossible".

**Comment by wing328** [comment]
Closing this as "allOf" is tracked via #2161

**Comment by demonfiddler** [comment]
Hi @wing328, I think we need to reopen this issue, as it is not addressed by the fix for Issue #2161.

**Comment by demonfiddler** [comment]
Closed, as PR#[2445](https://github.com/swagger-api/swagger-codegen/pull/2445) now merged.
